FI 1537 - Turn TCP Purple Errors to Red

Jack-Fraser16 opened the pull request:
Updated the HTTP and FHIR clients to do some error filtering. This should not affect errors raised by inability to connect to the validator(s), as those interactions do not use either the HTTP nor FHIR Client.

arscan commented on lib/inferno/dsl/fhir_client.rb:
```diff
@@ -50,15 +50,24 @@ def self.included(klass)
-          FHIRClientBuilder.new.build(self, self.class.fhir_client_definitions[client])
+        fhir_error_filter do
+          fhir_clients[client] ||=
+            FHIRClientBuilder.new.build(self, self.class.fhir_client_definitions[client])
```
This makes a network request?

arscan commented on lib/inferno/dsl/fhir_client.rb:
```diff
@@ -72,12 +81,14 @@ def fhir_clients
-          operation_headers = fhir_client(client).fhir_headers
-          operation_headers.merge!('Content-Type' => 'application/fhir+json') if body.present?
-          operation_headers.merge!(headers) if headers.present?
+        fhir_error_filter do
```
Ideally (though maybe its not practical), we would be able to document that we made the request and received no response, so it would show up in our interface.  By wrapping this outside of the store block, i assume we would still never store the request.  Maybe this would be a separate ticket though... I realize that this would be beyond the current scope.

arscan commented on spec/inferno/dsl/fhir_client_spec.rb:
```diff
@@ -230,6 +273,34 @@ def test_session_id
+          .and_raise(SocketError, 'Failed to open TCP')
+      end
+
+      it 'raises a skip' do
```
```suggestion
      it 'raises a test failure exception' do
```

arscan commented on lib/inferno/dsl/fhir_client.rb:
```diff
@@ -59,6 +59,13 @@ def fhir_clients
       end
 
+      # @private
+      def fhir_error_filter(&block)
```
I would rename this to: `fhir_client_tcp_exception_handler`  (and `http_client_tcp_exception_handler`).  Yeah its verbose, but might as well be clear about what it is doing.

Also, another option here would be to have a more generic `tcp_exception_handler` method, put that in another file (similar to request_storage... and require_relative/klass.include it in the same way in both clients), and then have that code have a 2 part rescue.  First, for `SocketErrors`, then a separate rescue for `Faraday:ConnectionFailed`.  That way you can make it so that method is a little less verbose.  One downside to this is that in practice that separate file will have to know how all clients work underneath the hood.  So I'm not that opinionated about it.

arscan commented:
I think you are missing a case because I'm still getting one purple error when i run the g10 tests here (or maybe this needs to get addressed in the US Core or g10 gems if they aren't using the http or fhir clients to make the /metadata request)?

![Screen Shot 2022-05-23 at 3 47 27 PM](https://user-images.githubusercontent.com/412901/169895097-9be61c7d-2d82-461d-97ad-41915a89ddd8.png)

Jack-Fraser16 commented:
> I think you are missing a case because I'm still getting one purple error when i run the g10 tests here (or maybe this needs to get addressed in the US Core or g10 gems if they aren't using the http or fhir clients to make the /metadata request)?
> 
> ![Screen Shot 2022-05-23 at 3 47 27 PM](https://user-images.githubusercontent.com/412901/169895097-9be61c7d-2d82-461d-97ad-41915a89ddd8.png)

This needs to be addressed in US Core kit. The call is coming from some `fhir_client.detect_version` which is either not built upon the inferno-core `fhir_client` or `detect_version` is a method defined within that kit. Either way, I'll add a ticket to investigate/fix.

Jack-Fraser16 commented:
Reworked the exception handler to live in its own module and included it in the necessary clients. Additionally, added the `tcp_exception_handler` to the `validate` call in `fhir_validation`.

arscan commented on lib/inferno/dsl/fhir_validation.rb:
```diff
@@ -1,4 +1,5 @@
 require_relative '../ext/fhir_models'
+require_relative 'tcp_exception_handler'
 
```
I think you can remove this now?
```suggestion

```

arscan commented on lib/inferno/dsl/fhir_validation.rb:
```diff
@@ -43,6 +44,8 @@ def find_validator(validator_name)
       end
 
       class Validator
+        include TCPExceptionHandler
+
```
And remove this too?

```suggestion
```
